core/src/block_sync.rs

```diff
@@ -25,66 +26,100 @@ pub struct BlockSynchronizer {
-    async fn handle(&mut self, message: message::Message) {
-        message.handle_message(self).await;
-    }
 }
 
 impl BlockSynchronizer {
-    /// Sends request for latest blocks to a chosen peer
-    async fn request_latest_blocks_from_peer(&mut self, peer_id: PeerId) {
-        message::Message::GetBlocksAfter(message::GetBlocksAfter::new(
-            self.sumeragi.latest_block_hash(),
-            self.peer_id.clone(),
-        ))
-        .send_to(self.broker.clone(), peer_id)
-        .await;
-    }
-
     /// Create [`Self`] from [`Configuration`]
     pub fn from_configuration(
         config: &Configuration,
         sumeragi: Arc<Sumeragi>,
+        p2p: Arc<P2PSystem>,
         peer_id: PeerId,
-        broker: Broker,
-    ) -> Self {
-        Self {
+    ) -> Arc<Self> {
+        let (incoming_message_sender, incoming_message_receiver) =
+            std::sync::mpsc::sync_channel(250);
+
+        Arc::new(Self {
             peer_id,
             sumeragi,
             gossip_period: Duration::from_millis(config.gossip_period_ms),
             block_batch_size: config.block_batch_size,
-            broker,
-            actor_channel_capacity: config.actor_channel_capacity,
+            p2p,
+            message_sender: Mutex::new(incoming_message_sender),
+            message_receiver: Mutex::new(incoming_message_receiver),
+        })
+    }
+
+    /// Deposit a block_sync network message.
+    #[allow(clippy::expect_used)]
+    pub fn incoming_message(&self, msg: message::VersionedMessage) {
+        println!("got msg");
+        if self
+            .message_sender
+            .lock()
+            .expect("Lock on sender")
+            .try_send(msg)
+            .is_err()
+        {
+            error!("This peer is faulty. Incoming messages have to be dropped due to low processing speed.");
+        }
+    }    
+}
+
+pub fn start_read_loop(block_sync: Arc<BlockSynchronizer>) -> ThreadHandler {
+    // Oneshot channel to allow forcefully stopping the thread.
+    let (shutdown_sender, shutdown_receiver) = tokio::sync::oneshot::channel();
+
+    let thread_handle = std::thread::spawn(move || {
+        block_sync_read_loop(&block_sync, shutdown_receiver);
+    });
+
+    let shutdown = move || {
+        let _result = shutdown_sender.send(());
+    };
+
+    ThreadHandler::new(Box::new(shutdown), thread_handle)
+}
+
+fn block_sync_read_loop(
+    block_sync: &BlockSynchronizer,
+    mut shutdown_receiver: tokio::sync::oneshot::Receiver<()>,
+) {
+    let mut message_receiver = block_sync.message_receiver.lock().unwrap();
+    let mut last_requested_blocks = Instant::now();
+    let mut request_blocks_peer_index = 0_usize;
+    loop {
+        // We have no obligations to network delivery so we simply exit on shutdown signal.
+        if shutdown_receiver.try_recv().is_ok() {
+            info!("P2P thread is being shut down");
+            return;
+        }
+        std::thread::sleep(Duration::from_millis(10));
```

Comment:
Sleep is necessary to yield the OS thread, though frankly the way we're doing it: rigidly, with arbitrary values, is error-prone. What happens if we have another problem on a weaker machine on which the sleep interval must be lower/higher? 

The only viable option I can think of at this hour is coroutines and callback functions. 

Given that we need `async` anyway, the implementation is likely going to be easier than it looks.

Reply:
> Sleep is necessary to yield the OS thread

Yes, i understand that, my concern is about taken architectural approach. 

We are burning cycles in a loop every 10ms when 1) shutdown is one time event 2) gossip period is 10000ms in default configuration and 3) and i highly doubt that we will receive messages from `message_reciever` with such frequency.

Tasks here (and in other places) seems to be about waiting a lot and there is mechanism to do this more effectively.

Reply:
> Yes, i understand that, my concern is about taken architectural approach.

Sorry, the reason I mentioned this, is because one of Iroha's targets is an OS-less deployment on embedded devices, which might not have a decent scheduler.

Reply:
> one of Iroha's targets is an OS-less deployment on embedded devices, which might not have a decent scheduler

Insteresting, it's our plans for the future to add support for such target?

Reply:
Yes. Long-term plan.